File: backend/app/models/download.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from enum import Enum
from datetime import datetime


class DownloadStatus(str, Enum):
    """下载状态枚举"""
    PENDING = "pending"
    DOWNLOADING = "downloading"
    PAUSED = "paused"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    ERROR = "error"


class DownloadSegment(BaseModel):
    """下载分段模型"""
    start: int
    end: int
    downloaded: int = 0
    status: str = DownloadStatus.PENDING


class DownloadRequest(BaseModel):
    """下载请求模型"""
    video_id: str
    filename: Optional[str] = None
    force: Optional[bool] = False


class DownloadAction(BaseModel):
    """下载操作模型"""
    video_id: str
    action: str  # 'pause', 'resume', 'cancel', 'retry', 'delete'


class DownloadProgress(BaseModel):
    """下载进度模型"""
    video_id: str
    filename: str
    title: Optional[str] = None
    cover_url: Optional[str] = None
    total_size: int
    downloaded: int
    status: str
    speed: float  # bytes per second
    error_message: Optional[str] = None
    url: str = ""
    created_at: datetime
    completed_at: Optional[datetime] = None
    retry_count: int = 0
    max_retries: int = 3
    segments: Optional[List[DownloadSegment]] = None


class DownloadConfig(BaseModel):
    """下载配置模型"""
    chunk_size: int = 1024 * 1024 * 16  # 16MB
    buffer_size: int = 1024 * 1024 * 4  # 4MB
    max_segments: int = 4
    min_segment_size: int = 40 * 1024 * 1024  # 40MB
    max_retries: int = 3
    timeout: float = 30.0
    progress_update_interval: float = 0.5 

File: frontend/src/types/download.ts
/**
 * 下载相关类型定义
 */

// 下载操作类型
export type DownloadActionType = 'pause' | 'resume' | 'cancel' | 'retry' | 'delete';

// 下载状态类型
export type DownloadStatus = 'pending' | 'downloading' | 'paused' | 'completed' | 'cancelled' | 'error';

// 下载分段类型
export interface DownloadSegment {
  start: number;
  end: number;
  downloaded: number;
  status: string;
}

// 下载进度类型
export interface DownloadProgress {
  video_id: string;
  filename: string;
  title?: string;
  cover_url?: string;
  total_size: number;
  downloaded: number;
  status: DownloadStatus;
  speed: number;
  error_message?: string;
  url: string;
  created_at: string;
  completed_at?: string;
  retry_count: number;
  max_retries: number;
  segments?: DownloadSegment[];
}

// 下载请求类型
export interface DownloadRequest {
  video_id: string;
  filename?: string;
  force?: boolean;
}

// 下载操作请求类型
export interface DownloadAction {
  video_id: string;
  action: DownloadActionType;
}

// 下载操作响应类型
export interface DownloadResponse {
  status: 'success' | 'error' | 'warning';
  message?: string;
  video_id?: string;
  existing_download?: DownloadProgress;
} 